### PR TITLE
vivacity note feature

### PR DESCRIPTION
**Pull request by bricklayerrrr**
To improve the UX of Vivacity, added the feature for users to interact with NOTE instead of cNOTE
VivaRouter converts Note to cNOTE and do all the action on behalf of user.

VCNote has added a function that specifies the address to receive borrowed assets and a permit function based on a signature.

VCNote:
function that specifies the address to receive borrowed assets:
1. borrow(uint borrowAmount, address payable receiver): 
2. borrowInternal(uint borrowAmount, address payable receiver)
3. borrowFresh(address payable borrower, uint borrowAmount, address payable receiver)
* If parameters and calls are excluded, this is the only change in the "doTransferOut" code in browFresh.

permit function based on a signature:
1. borrowPermit(BorrowPermitParams memory params)
2. borrowPermitInternal(BorrowPermitParams memory params) 
3. BorrowPermitParams Libarary

additional:
change vcNote's storage to make it easier to upgrade

### CODE REVIEW

File: contracts/vcnote/storages/VCNoteStorage.sol [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.10;
+
+struct VCNoteStorage {
+    address lendingLedger;                        // lending ledger address
+    mapping(address => uint256) nonces;           // nonces for permit
```
Comment (solidity-github) [inline contracts/vcnote/storages/VCNoteStorage.sol:6]: What is nonce used for?

**Review by solidity-github — COMMENTED** [review]
Is there a reason to use contract instead of interface?

File: contracts/vcnote/VCNoteRouter.sol [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        CErc20(s.cNOTE).mint(amount);
+
+        // mint vcNOTE in vivacity
+        uint256 mintedCNote = IERC20(s.cNOTE).balanceOf(address(this)) - balanceCNoteBefore;
+        CErc20(s.vcNOTE).mint(mintedCNote);
+
+        // transfer vcNOTE to the user
+        uint256 mintedVCNote = IERC20(s.vcNOTE).balanceOf(address(this)) - balanceVCNoteBefore;
+        IERC20(s.vcNOTE).transfer(msg.sender, mintedVCNote);
+    }
```
Comment (jeonghoyeo7) [inline contracts/vcnote/VCNoteRouter.sol:38]: What about emitting an event for this mint (also as redeem/repay as well below)?

File: contracts/vcnote/VCNoteRouter.sol [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+        // repayBorrow in clm
+        uint256 mintedCNote = IERC20(s.cNOTE).balanceOf(address(this)) - balanceCNoteBefore;
+        CErc20(s.vcNOTE).repayBorrowBehalf(msg.sender, mintedCNote);
+    }
+
+    function borrow(BorrowPermitParams memory params) external {
+        VCNoteRouterStorage memory s = VCNoteRouterStorageLib.get();
+
+        uint256 balanceBefore = IERC20(s.cNOTE).balanceOf(address(this));
```
Comment (jeonghoyeo7) [inline contracts/vcnote/VCNoteRouter.sol:79]: suggest to keep consistency: from `balanceBefore` to `balanceCNoteBefore`

File: contracts/vcnote/VCNoteRouter.sol [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+        uint256 balanceAfter = IERC20(s.cNOTE).balanceOf(address(this));
+        IERC20(s.cNOTE).transfer(msg.sender, balanceAfter - balanceBefore);
+    }
+
+    function liquidateBorrow(address borrower, uint repayAmount, CTokenInterface cTokenCollateral) external {
+        VCNoteRouterStorage memory s = VCNoteRouterStorageLib.get();
+
+        uint256 balanceCNoteBefore = IERC20(s.cNOTE).balanceOf(address(this));
+        uint256 balanceCollateralBefore = IERC20(address(cTokenCollateral)).balanceOf(address(this));
```
Comment (solidity-github) [inline contracts/vcnote/VCNoteRouter.sol:91]: Why cast type to IERC20..?

Comment (solidity-github) [reply contracts/vcnote/VCNoteRouter.sol:91]: Isn't there balanceOf on CtokenInterface?

File: contracts/vcnote/VCNoteRouter.sol [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        IERC20(s.NOTE).transferFrom(msg.sender, address(this), repayAmount);
+
+        CErc20(s.cNOTE).mint(repayAmount);
+
+        uint256 mintedCNote = IERC20(s.cNOTE).balanceOf(address(this)) - balanceCNoteBefore;
+
+        VCNote(s.vcNOTE).liquidateBorrow(borrower, mintedCNote, cTokenCollateral);
+
+        uint256 balanceCollateralAfter = IERC20(address(cTokenCollateral)).balanceOf(address(this));
+        IERC20(address(cTokenCollateral)).transfer(msg.sender, balanceCollateralAfter - balanceCollateralBefore);
```
Comment (solidity-github) [inline contracts/vcnote/VCNoteRouter.sol:102]: Same as the above review.

File: contracts/vcnote/VCNoteRouter.sol [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    function borrow(BorrowPermitParams memory params) external {
+        VCNoteRouterStorage memory s = VCNoteRouterStorageLib.get();
+
+        uint256 balanceBefore = IERC20(s.cNOTE).balanceOf(address(this));
+        // borrow in clm
+        VCNote(s.vcNOTE).borrowPermit(params);
+
+        uint256 balanceAfter = IERC20(s.cNOTE).balanceOf(address(this));
+        IERC20(s.cNOTE).transfer(msg.sender, balanceAfter - balanceBefore);
```
Comment (solidity-github) [inline contracts/vcnote/VCNoteRouter.sol:84]: Is there a cNote to transfer?

**Pull request merged**